From: Jean Delvare <jdelvare@suse.de>
Subject: ACPI, APEI: Fixup common access width firmware bug
References: fate#311701
Patch-Mainline: v3.6-rc2
Git-commit: f712c71f7b2b43b894d1e92e1b77385fcad8815f


Signed-off-by: Thomas Renninger <trenn@suse.de>

Many firmwares have a common register definition bug where 8-bit
access width is specified for a 32-bit register. Ideally this should
be fixed in the BIOS, but earlier versions of the kernel did not
complain, so fix that up silently.

This closes kernel bug #43282:
https://bugzilla.kernel.org/show_bug.cgi?id=43282

Signed-off-by: Jean Delvare <jdelvare@suse.de>
Acked-by: Huang Ying <ying.huang@intel.com>
Acked-by: Gary Hade <garyhade@us.ibm.com>
Cc: stable@vger.kernel.org [3.4+]
Signed-off-by: Len Brown <len.brown@intel.com>

---
 drivers/acpi/apei/apei-base.c |    5 +++++
 1 file changed, 5 insertions(+)

Index: linux-3.0-SLE11-SP3/drivers/acpi/apei/apei-base.c
===================================================================
--- linux-3.0-SLE11-SP3.orig/drivers/acpi/apei/apei-base.c
+++ linux-3.0-SLE11-SP3/drivers/acpi/apei/apei-base.c
@@ -548,6 +548,11 @@ static int apei_check_gar(struct acpi_ge
 	}
 	*access_bit_width = 1UL << (access_size_code + 2);
 
+	/* Fixup common BIOS bug */
+	if (bit_width == 32 && bit_offset == 0 && (*paddr & 0x03) == 0 &&
+	    *access_bit_width < 32)
+		*access_bit_width = 32;
+
 	if ((bit_width + bit_offset) > *access_bit_width) {
 		pr_warning(FW_BUG APEI_PFX
 			   "Invalid bit width + offset in GAR [0x%llx/%u/%u/%u/%u]\n",
